DB transaction loops forever with an exception in new leader

I have seen the following error a couple of times in docker-compose setup, after leadership change occurs. The transaction retries forever because there is an exception. It doesn't block the processor. It can still create workflows and activate jobs.
`Caused by: java.lang.IllegalStateException: Expected to find job with key 2251799813692390, but no job found`
I wonder if it is due to some inconsistent state.

```
java.lang.RuntimeException: Unexpected error occurred during zeebe db transaction operation.
broker-2_1  | 	at io.zeebe.db.impl.rocksdb.transaction.DefaultDbContext.runInTransaction(DefaultDbContext.java:155) ~[zeebe-db-0.18.0-SNAPSHOT.jar:0.18.0-SNAPSHOT]
broker-2_1  | 	at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.ensureInOpenTransaction(ZeebeTransactionDb.java:154) ~[zeebe-db-0.18.0-SNAPSHOT.jar:0.18.0-SNAPSHOT]
broker-2_1  | 	at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.whileTrue(ZeebeTransactionDb.java:266) ~[zeebe-db-0.18.0-SNAPSHOT.jar:0.18.0-SNAPSHOT]
broker-2_1  | 	at io.zeebe.db.impl.rocksdb.transaction.TransactionalColumnFamily.whileTrue(TransactionalColumnFamily.java:117) ~[zeebe-db-0.18.0-SNAPSHOT.jar:0.18.0-SNAPSHOT]
broker-2_1  | 	at io.zeebe.db.impl.rocksdb.transaction.TransactionalColumnFamily.whileTrue(TransactionalColumnFamily.java:108) ~[zeebe-db-0.18.0-SNAPSHOT.jar:0.18.0-SNAPSHOT]
broker-2_1  | 	at io.zeebe.db.impl.rocksdb.transaction.TransactionalColumnFamily.whileTrue(TransactionalColumnFamily.java:104) ~[zeebe-db-0.18.0-SNAPSHOT.jar:0.18.0-SNAPSHOT]
broker-2_1  | 	at io.zeebe.engine.state.instance.JobState.forEachTimedOutEntry(JobState.java:176) ~[zeebe-workflow-engine-0.18.0-SNAPSHOT.jar:0.18.0-SNAPSHOT]
broker-2_1  | 	at io.zeebe.engine.processor.workflow.job.JobTimeoutTrigger.deactivateTimedOutJobs(JobTimeoutTrigger.java:61) ~[zeebe-workflow-engine-0.18.0-SNAPSHOT.jar:0.18.0-SNAPSHOT]
broker-2_1  | 	at io.zeebe.util.sched.ActorJob.invoke(ActorJob.java:90) ~[zeebe-util-0.18.0-SNAPSHOT.jar:0.18.0-SNAPSHOT]
broker-2_1  | 	at io.zeebe.util.sched.ActorJob.execute(ActorJob.java:53) [zeebe-util-0.18.0-SNAPSHOT.jar:0.18.0-SNAPSHOT]
broker-2_1  | 	at io.zeebe.util.sched.ActorTask.execute(ActorTask.java:189) [zeebe-util-0.18.0-SNAPSHOT.jar:0.18.0-SNAPSHOT]
broker-2_1  | 	at io.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:154) [zeebe-util-0.18.0-SNAPSHOT.jar:0.18.0-SNAPSHOT]
broker-2_1  | 	at io.zeebe.util.sched.ActorThread.doWork(ActorThread.java:135) [zeebe-util-0.18.0-SNAPSHOT.jar:0.18.0-SNAPSHOT]
broker-2_1  | 	at io.zeebe.util.sched.ActorThread.run(ActorThread.java:112) [zeebe-util-0.18.0-SNAPSHOT.jar:0.18.0-SNAPSHOT]
broker-2_1  | Caused by: java.lang.IllegalStateException: Expected to find job with key 2251799813692390, but no job found
broker-2_1  | 	at io.zeebe.engine.state.instance.JobState.visitJob(JobState.java:225) ~[zeebe-workflow-engine-0.18.0-SNAPSHOT.jar:0.18.0-SNAPSHOT]
broker-2_1  | 	at io.zeebe.engine.state.instance.JobState.lambda$forEachTimedOutEntry$0(JobState.java:182) ~[zeebe-workflow-engine-0.18.0-SNAPSHOT.jar:0.18.0-SNAPSHOT]
broker-2_1  | 	at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.visit(ZeebeTransactionDb.java:365) ~[zeebe-db-0.18.0-SNAPSHOT.jar:0.18.0-SNAPSHOT]
broker-2_1  | 	at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.lambda$whileTrue$8(ZeebeTransactionDb.java:273) ~[zeebe-db-0.18.0-SNAPSHOT.jar:0.18.0-SNAPSHOT]
broker-2_1  | 	at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.lambda$ensureInOpenTransaction$1(ZeebeTransactionDb.java:155) ~[zeebe-db-0.18.0-SNAPSHOT.jar:0.18.0-SNAPSHOT]
broker-2_1  | 	at io.zeebe.db.impl.rocksdb.transaction.DefaultDbContext.runInNewTransaction(DefaultDbContext.java:171) ~[zeebe-db-0.18.0-SNAPSHOT.jar:0.18.0-SNAPSHOT]
broker-2_1  | 	at io.zeebe.db.impl.rocksdb.transaction.DefaultDbContext.runInTransaction(DefaultDbContext.java:143) ~[zeebe-db-0.18.0-SNAPSHOT.jar:0.18.0-SNAPSHOT]
```

@npepinpe I think you are right since the transaction is then aborted and rolled back

See also the following traces which we will print every 30 seconds the same key:

```
12:58:23.982 [stream-processor] [zeebe-0.zeebe.default.svc.cluster.local:26501-zb-actors-2] ERROR io.zeebe.util.actor - Actor failed in phase 'STARTED'. Continue with next job.
java.lang.RuntimeException: Unexpected error occurred during zeebe db transaction operation.
        at io.zeebe.db.impl.rocksdb.transaction.DefaultDbContext.runInTransaction(DefaultDbContext.java:155) ~[zeebe-db-0.19.0-SNAPSHOT.jar:0.19.0-SNAPSHOT]
        at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.ensureInOpenTransaction(ZeebeTransactionDb.java:154) ~[zeebe-db-0.19.0-SNAPSHOT.jar:0.19.0-SNAPSHOT]
        at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.whileTrue(ZeebeTransactionDb.java:266) ~[zeebe-db-0.19.0-SNAPSHOT.jar:0.19.0-SNAPSHOT]
        at io.zeebe.db.impl.rocksdb.transaction.TransactionalColumnFamily.whileTrue(TransactionalColumnFamily.java:117) ~[zeebe-db-0.19.0-SNAPSHOT.jar:0.19.0-SNAPSHOT]
        at io.zeebe.db.impl.rocksdb.transaction.TransactionalColumnFamily.whileTrue(TransactionalColumnFamily.java:108) ~[zeebe-db-0.19.0-SNAPSHOT.jar:0.19.0-SNAPSHOT]
        at io.zeebe.db.impl.rocksdb.transaction.TransactionalColumnFamily.whileTrue(TransactionalColumnFamily.java:104) ~[zeebe-db-0.19.0-SNAPSHOT.jar:0.19.0-SNAPSHOT]
        at io.zeebe.engine.state.instance.JobState.forEachTimedOutEntry(JobState.java:198) ~[zeebe-workflow-engine-0.19.0-SNAPSHOT.jar:0.19.0-SNAPSHOT]
        at io.zeebe.engine.processor.workflow.job.JobTimeoutTrigger.deactivateTimedOutJobs(JobTimeoutTrigger.java:61) ~[zeebe-workflow-engine-0.19.0-SNAPSHOT.jar:0.19.0-SNAPSHOT]
        at io.zeebe.util.sched.ActorJob.invoke(ActorJob.java:88) ~[zeebe-util-0.19.0-SNAPSHOT.jar:0.19.0-SNAPSHOT]
        at io.zeebe.util.sched.ActorJob.execute(ActorJob.java:51) [zeebe-util-0.19.0-SNAPSHOT.jar:0.19.0-SNAPSHOT]
        at io.zeebe.util.sched.ActorTask.execute(ActorTask.java:177) [zeebe-util-0.19.0-SNAPSHOT.jar:0.19.0-SNAPSHOT]
        at io.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:142) [zeebe-util-0.19.0-SNAPSHOT.jar:0.19.0-SNAPSHOT]
        at io.zeebe.util.sched.ActorThread.doWork(ActorThread.java:126) [zeebe-util-0.19.0-SNAPSHOT.jar:0.19.0-SNAPSHOT]
        at io.zeebe.util.sched.ActorThread.run(ActorThread.java:104) [zeebe-util-0.19.0-SNAPSHOT.jar:0.19.0-SNAPSHOT]
Caused by: java.lang.IllegalStateException: Expected to find job with key 9007199254773553, but no job found
        at io.zeebe.engine.state.instance.JobState.visitJob(JobState.java:247) ~[zeebe-workflow-engine-0.19.0-SNAPSHOT.jar:0.19.0-SNAPSHOT]
        at io.zeebe.engine.state.instance.JobState.lambda$forEachTimedOutEntry$0(JobState.java:204) ~[zeebe-workflow-engine-0.19.0-SNAPSHOT.jar:0.19.0-SNAPSHOT]
        at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.visit(ZeebeTransactionDb.java:365) ~[zeebe-db-0.19.0-SNAPSHOT.jar:0.19.0-SNAPSHOT]
        at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.lambda$whileTrue$8(ZeebeTransactionDb.java:273) ~[zeebe-db-0.19.0-SNAPSHOT.jar:0.19.0-SNAPSHOT]
        at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.lambda$ensureInOpenTransaction$1(ZeebeTransactionDb.java:155) ~[zeebe-db-0.19.0-SNAPSHOT.jar:0.19.0-SNAPSHOT]
        at io.zeebe.db.impl.rocksdb.transaction.DefaultDbContext.runInTransaction(DefaultDbContext.java:141) ~[zeebe-db-0.19.0-SNAPSHOT.jar:0.19.0-SNAPSHOT]
        ... 13 more
12:58:53.990 [stream-processor] [zeebe-0.zeebe.default.svc.cluster.local:26501-zb-actors-2] ERROR io.zeebe.util.actor - Actor failed in phase 'STARTED'. Continue with next job.
java.lang.RuntimeException: Unexpected error occurred during zeebe db transaction operation.
        at io.zeebe.db.impl.rocksdb.transaction.DefaultDbContext.runInTransaction(DefaultDbContext.java:155) ~[zeebe-db-0.19.0-SNAPSHOT.jar:0.19.0-SNAPSHOT]
        at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.ensureInOpenTransaction(ZeebeTransactionDb.java:154) ~[zeebe-db-0.19.0-SNAPSHOT.jar:0.19.0-SNAPSHOT]
        at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.whileTrue(ZeebeTransactionDb.java:266) ~[zeebe-db-0.19.0-SNAPSHOT.jar:0.19.0-SNAPSHOT]
        at io.zeebe.db.impl.rocksdb.transaction.TransactionalColumnFamily.whileTrue(TransactionalColumnFamily.java:117) ~[zeebe-db-0.19.0-SNAPSHOT.jar:0.19.0-SNAPSHOT]
        at io.zeebe.db.impl.rocksdb.transaction.TransactionalColumnFamily.whileTrue(TransactionalColumnFamily.java:108) ~[zeebe-db-0.19.0-SNAPSHOT.jar:0.19.0-SNAPSHOT]
        at io.zeebe.db.impl.rocksdb.transaction.TransactionalColumnFamily.whileTrue(TransactionalColumnFamily.java:104) ~[zeebe-db-0.19.0-SNAPSHOT.jar:0.19.0-SNAPSHOT]
        at io.zeebe.engine.state.instance.JobState.forEachTimedOutEntry(JobState.java:198) ~[zeebe-workflow-engine-0.19.0-SNAPSHOT.jar:0.19.0-SNAPSHOT]
        at io.zeebe.engine.processor.workflow.job.JobTimeoutTrigger.deactivateTimedOutJobs(JobTimeoutTrigger.java:61) ~[zeebe-workflow-engine-0.19.0-SNAPSHOT.jar:0.19.0-SNAPSHOT]
        at io.zeebe.util.sched.ActorJob.invoke(ActorJob.java:88) ~[zeebe-util-0.19.0-SNAPSHOT.jar:0.19.0-SNAPSHOT]
        at io.zeebe.util.sched.ActorJob.execute(ActorJob.java:51) [zeebe-util-0.19.0-SNAPSHOT.jar:0.19.0-SNAPSHOT]
        at io.zeebe.util.sched.ActorTask.execute(ActorTask.java:177) [zeebe-util-0.19.0-SNAPSHOT.jar:0.19.0-SNAPSHOT]
        at io.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:142) [zeebe-util-0.19.0-SNAPSHOT.jar:0.19.0-SNAPSHOT]
        at io.zeebe.util.sched.ActorThread.doWork(ActorThread.java:126) [zeebe-util-0.19.0-SNAPSHOT.jar:0.19.0-SNAPSHOT]
        at io.zeebe.util.sched.ActorThread.run(ActorThread.java:104) [zeebe-util-0.19.0-SNAPSHOT.jar:0.19.0-SNAPSHOT]
Caused by: java.lang.IllegalStateException: Expected to find job with key 9007199254773553, but no job found
        at io.zeebe.engine.state.instance.JobState.visitJob(JobState.java:247) ~[zeebe-workflow-engine-0.19.0-SNAPSHOT.jar:0.19.0-SNAPSHOT]
        at io.zeebe.engine.state.instance.JobState.lambda$forEachTimedOutEntry$0(JobState.java:204) ~[zeebe-workflow-engine-0.19.0-SNAPSHOT.jar:0.19.0-SNAPSHOT]
        at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.visit(ZeebeTransactionDb.java:365) ~[zeebe-db-0.19.0-SNAPSHOT.jar:0.19.0-SNAPSHOT]
        at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.lambda$whileTrue$8(ZeebeTransactionDb.java:273) ~[zeebe-db-0.19.0-SNAPSHOT.jar:0.19.0-SNAPSHOT]
        at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.lambda$ensureInOpenTransaction$1(ZeebeTransactionDb.java:155) ~[zeebe-db-0.19.0-SNAPSHOT.jar:0.19.0-SNAPSHOT]
        at io.zeebe.db.impl.rocksdb.transaction.DefaultDbContext.runInNewTransaction(DefaultDbContext.java:171) ~[zeebe-db-0.19.0-SNAPSHOT.jar:0.19.0-SNAPSHOT]
        at io.zeebe.db.impl.rocksdb.transaction.DefaultDbContext.runInTransaction(DefaultDbContext.java:143) ~[zeebe-db-0.19.0-SNAPSHOT.jar:0.19.0-SNAPSHOT]
        ... 13 more
```

--- a/engine/src/main/java/io/zeebe/engine/processor/workflow/job/JobBatchActivateProcessor.java
+++ b/engine/src/main/java/io/zeebe/engine/processor/workflow/job/JobBatchActivateProcessor.java
@@ -17,8 +17,6 @@
  */
 package io.zeebe.engine.processor.workflow.job;
 
-import static io.zeebe.util.sched.clock.ActorClock.currentTimeMillis;
-
 import io.zeebe.engine.processor.KeyGenerator;
 import io.zeebe.engine.processor.TypedRecord;
 import io.zeebe.engine.processor.TypedRecordProcessor;
@@ -120,7 +118,7 @@ public class JobBatchActivateProcessor implements TypedRecordProcessor<JobBatchR
         value.getTypeBuffer(),
         (key, jobRecord) -> {
           int remainingAmount = amount.get();
-          final long deadline = currentTimeMillis() + value.getTimeout();
+          final long deadline = record.getTimestamp() + value.getTimeout();
           jobRecord.setDeadline(deadline).setWorker(value.getWorkerBuffer());
 
           // fetch and set variables, required here to already have the full size of the job record

--- a/engine/src/test/java/io/zeebe/engine/processor/workflow/job/ActivateJobsTest.java
+++ b/engine/src/test/java/io/zeebe/engine/processor/workflow/job/ActivateJobsTest.java
@@ -19,7 +19,6 @@ package io.zeebe.engine.processor.workflow.job;
 
 import static io.zeebe.protocol.record.Assertions.assertThat;
 import static io.zeebe.test.util.TestUtil.waitUntil;
-import static io.zeebe.test.util.record.RecordingExporter.jobBatchRecords;
 import static io.zeebe.test.util.record.RecordingExporter.jobRecords;
 import static io.zeebe.test.util.record.RecordingExporter.workflowInstanceRecords;
 import static org.assertj.core.api.Assertions.assertThat;
@@ -343,7 +342,6 @@ public class ActivateJobsTest {
 
     final String worker = "testWorker";
     final Duration timeout = Duration.ofMinutes(4);
-    final long deadline = clock.getCurrentTime().plus(timeout).toEpochMilli();
 
     ENGINE.deployment().withXmlResource(PROCESS_ID, MODEL_SUPPLIER.apply(taskType)).deploy();
     createWorkflowInstances(1, "{'foo':'bar'}");
@@ -351,43 +349,42 @@ public class ActivateJobsTest {
         jobRecords(JobIntent.CREATED).withType(taskType).getFirst();
 
     // when
-    final long jobKey =
+    final Record<JobBatchRecordValue> jobActivatedRecord =
         ENGINE
             .jobs()
             .withType(taskType)
             .byWorker(worker)
             .withTimeout(timeout.toMillis())
             .withMaxJobsToActivate(1)
-            .activate()
-            .getKey();
+            .activate();
 
-    final JobRecordValue job =
-        jobBatchRecords(JobBatchIntent.ACTIVATED)
-            .withRecordKey(jobKey)
-            .getFirst()
-            .getValue()
-            .getJobs()
-            .get(0);
+    final JobRecordValue jobActivated = jobActivatedRecord.getValue().getJobs().get(0);
+    final Record<JobBatchRecordValue> jobActivate =
+        RecordingExporter.jobBatchRecords()
+            .withType(taskType)
+            .withIntent(JobBatchIntent.ACTIVATE)
+            .getFirst();
 
     // then
-    Assertions.assertThat(job)
+    Assertions.assertThat(jobActivated)
         .hasType(taskType)
         .hasWorker(worker)
         .hasRetries(3)
-        .hasDeadline(deadline);
+        .hasDeadline(jobActivate.getTimestamp() + timeout.toMillis());
 
-    assertThat(job.getVariables()).containsExactly(entry("foo", "bar"));
+    assertThat(jobActivated.getVariables()).containsExactly(entry("foo", "bar"));
 
     final JobRecordValue jobRecordValue = jobRecord.getValue();
-    assertThat(job.getWorkflowInstanceKey()).isEqualTo(jobRecordValue.getWorkflowInstanceKey());
-    Assertions.assertThat(job)
+    assertThat(jobActivated.getWorkflowInstanceKey())
+        .isEqualTo(jobRecordValue.getWorkflowInstanceKey());
+    Assertions.assertThat(jobActivated)
         .hasBpmnProcessId(jobRecordValue.getBpmnProcessId())
         .hasWorkflowDefinitionVersion(jobRecordValue.getWorkflowDefinitionVersion())
         .hasWorkflowKey(jobRecordValue.getWorkflowKey())
         .hasElementId(jobRecordValue.getElementId())
         .hasElementInstanceKey(jobRecordValue.getElementInstanceKey());
 
-    assertThat(job.getCustomHeaders()).isEqualTo(jobRecordValue.getCustomHeaders());
+    assertThat(jobActivated.getCustomHeaders()).isEqualTo(jobRecordValue.getCustomHeaders());
   }
 
   @Test

--- a/engine/src/test/java/io/zeebe/engine/processor/workflow/job/JobTimeOutTest.java
+++ b/engine/src/test/java/io/zeebe/engine/processor/workflow/job/JobTimeOutTest.java
@@ -63,12 +63,12 @@ public class JobTimeOutTest {
     ENGINE.increaseTime(JobTimeoutTrigger.TIME_OUT_POLLING_INTERVAL);
 
     // when expired
-    RecordingExporter.jobRecords(TIME_OUT).getFirst();
+    RecordingExporter.jobRecords(TIME_OUT).withType(jobType).getFirst();
     ENGINE.jobs().withType(jobType).activate();
 
     // then activated again
     final List<Record<JobRecordValue>> jobEvents =
-        RecordingExporter.jobRecords().limit(6).collect(Collectors.toList());
+        RecordingExporter.jobRecords().withType(jobType).limit(6).collect(Collectors.toList());
 
     assertThat(jobEvents).extracting(Record::getKey).contains(jobKey);
     assertThat(jobEvents)
@@ -80,6 +80,29 @@ public class JobTimeOutTest {
             JobIntent.TIME_OUT,
             JobIntent.TIMED_OUT,
             JobIntent.ACTIVATED);
+  }
+
+  @Test
+  public void shouldTimeOutAfterReprocessing() {
+    // given
+    final long jobKey = ENGINE.createJob(jobType, PROCESS_ID).getKey();
+    final long timeout = 10L;
+
+    ENGINE.jobs().withType(jobType).withTimeout(timeout).activate();
+    ENGINE.increaseTime(JobTimeoutTrigger.TIME_OUT_POLLING_INTERVAL);
+    jobRecords(TIME_OUT).withRecordKey(jobKey).getFirst();
+
+    final long jobKey2 = ENGINE.createJob(jobType, PROCESS_ID).getKey();
+    ENGINE.jobs().withType(jobType).activate();
+    ENGINE.job().withKey(jobKey).complete();
+
+    // when
+    ENGINE.reprocess();
+    ENGINE.jobs().withType(jobType).activate();
+
+    // then
+    ENGINE.increaseTime(JobTimeoutTrigger.TIME_OUT_POLLING_INTERVAL);
+    jobRecords(TIME_OUT).withRecordKey(jobKey2).getFirst();
   }
 
   @Test

--- a/engine/src/test/java/io/zeebe/engine/util/EngineRule.java
+++ b/engine/src/test/java/io/zeebe/engine/util/EngineRule.java
@@ -49,8 +49,10 @@ import io.zeebe.protocol.record.Record;
 import io.zeebe.protocol.record.intent.DeploymentIntent;
 import io.zeebe.protocol.record.intent.JobIntent;
 import io.zeebe.protocol.record.value.JobRecordValue;
+import io.zeebe.test.util.TestUtil;
 import io.zeebe.test.util.record.RecordingExporter;
 import io.zeebe.test.util.record.RecordingExporterTestWatcher;
+import io.zeebe.util.FileUtil;
 import io.zeebe.util.buffer.BufferWriter;
 import io.zeebe.util.sched.ActorCondition;
 import io.zeebe.util.sched.ActorControl;
@@ -102,7 +104,10 @@ public class EngineRule extends ExternalResource {
 
   @Override
   protected void before() {
+    startProcessors();
+  }
 
+  private void startProcessors() {
     final DeploymentRecord deploymentRecord = new DeploymentRecord();
     final UnsafeBuffer deploymentBuffer = new UnsafeBuffer(new byte[deploymentRecord.getLength()]);
     deploymentRecord.write(deploymentBuffer, 0);
@@ -137,6 +142,29 @@ public class EngineRule extends ExternalResource {
 
   public void increaseTime(Duration duration) {
     environmentRule.getClock().addTime(duration);
+  }
+
+  public void reprocess() {
+    forEachPartition(
+        partitionId -> {
+          try {
+            environmentRule.closeStreamProcessor(partitionId);
+            FileUtil.deleteFolder(
+                environmentRule
+                    .getStateSnapshotController(partitionId)
+                    .getLastValidSnapshotDirectory()
+                    .toPath());
+          } catch (Exception e) {
+            throw new RuntimeException(e);
+          }
+        });
+
+    final int lastSize = RecordingExporter.getRecords().size();
+    // we need to reset the record exporter
+    RecordingExporter.reset();
+
+    startProcessors();
+    TestUtil.waitUntil(() -> RecordingExporter.getRecords().size() >= lastSize);
   }
 
   public List<Integer> getPartitionIds() {

--- a/engine/src/test/java/io/zeebe/engine/util/MockTypedRecord.java
+++ b/engine/src/test/java/io/zeebe/engine/util/MockTypedRecord.java
@@ -31,13 +31,13 @@ public class MockTypedRecord<T extends UnifiedRecordValue> implements TypedRecor
   private long key;
   private RecordMetadata metadata;
   private T value;
-
-  public MockTypedRecord() {}
+  private final long timestamp;
 
   public MockTypedRecord(long key, RecordMetadata metadata, T value) {
     this.key = key;
     this.metadata = metadata;
     this.value = value;
+    this.timestamp = System.currentTimeMillis();
   }
 
   @Override
@@ -84,7 +84,7 @@ public class MockTypedRecord<T extends UnifiedRecordValue> implements TypedRecor
 
   @Override
   public long getTimestamp() {
-    throw new UnsupportedOperationException("not yet implemented");
+    return timestamp;
   }
 
   @Override

--- a/engine/src/test/java/io/zeebe/engine/util/StreamProcessorRule.java
+++ b/engine/src/test/java/io/zeebe/engine/util/StreamProcessorRule.java
@@ -130,16 +130,24 @@ public class StreamProcessorRule implements TestRule {
         }));
   }
 
+  public void closeStreamProcessor(int partitionId) throws Exception {
+    streams.closeProcessor(getLogName(partitionId));
+  }
+
   public void closeStreamProcessor() throws Exception {
-    streams.closeProcessor(getLogName(startPartitionId));
+    closeStreamProcessor(startPartitionId);
   }
 
   public long getCommitPosition() {
     return streams.getLogStream(getLogName(startPartitionId)).getCommitPosition();
   }
 
+  public StateSnapshotController getStateSnapshotController(int partitionId) {
+    return streams.getStateSnapshotController(getLogName(partitionId));
+  }
+
   public StateSnapshotController getStateSnapshotController() {
-    return streams.getStateSnapshotController(getLogName(startPartitionId));
+    return getStateSnapshotController(startPartitionId);
   }
 
   public CommandResponseWriter getCommandResponseWriter() {

--- a/engine/src/test/java/io/zeebe/engine/util/TestStreams.java
+++ b/engine/src/test/java/io/zeebe/engine/util/TestStreams.java
@@ -263,6 +263,11 @@ public class TestStreams {
 
     final ActorFuture<Void> openFuture = new CompletableActorFuture<>();
 
+    try {
+      currentSnapshotController.recover();
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
     zeebeDb = currentSnapshotController.openDb();
     final StreamProcessor processorService =
         StreamProcessor.builder()

